nixos/modules/services/matrix/synapse.nix

```diff
@@ -189,6 +191,18 @@ in {
         '';
       };
 
+      overrideLocalPostgresCheck = mkOption {
```

Comment:
I think that the naming is slightly misleading since the condition using this option (`hasLocalPostgresDB`) also influences how unit ordering of `matrix-synapse.service` looks like (i.e. whether or not to depend on `postgresql.service`) and the option's name and description should reflect that.

Unfortunately I can't think of a better option name right now, sorry.

Reply:
This is a good point, and something I didn't think of. However, I find it surprising / magic(TM) that this is at all dependent on the database URL as configured in the synapse configuration.

It seems that currently there is no strict dependency on PostgreSQL other than both being enabled, and the start order being encoded in the synapse unit's `After=` setting. In contrast to `Requires=`, the [`systemd.unit` manpage](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Before=) seems to suggest that this solely ensures ordering, but does not specify a hard requirement on the specified unit. This also holds in practice: specifying a non-existent service in `After=` is fine, whereas `Requires=` will generate a warning.

Keeping this in mind, it seems there's two options here:
- Right now, a local PostgreSQL database configured within Synapse will require PostgreSQL to be enabled, and started beforehand. We can generalize this behavior and say, if a PostgreSQL service is enabled through the corresponding `NixOS` module (generating a `postgres.service` unit), always start synapse after that (unconditionally add `After=postgres.service`). The one edge-case where the behavior would be different is when the system hosts a PostgreSQL database, but Synapse connects to a non-local instance instead.
- When a user chooses to override this check, it doesn't really make any sense to depend on an arbitrary `postgres.service`, either specified manually or generated through the corresponding unit file. Thus it also seems perfectly reasonable to remove this `After=` dependency when the override is used.

I also can't really come up with a better name. I'm leaning towards the latter option; could a compromise be that I add this as a remark in the option's documentation? Given that this is an edge-case, it seems fair to assume users would read that comment.

Reply:
> However, I find it surprising / magic(TM) that this is at all dependent on the database URL as configured in the synapse configuration

Yeah, noticed that too and wanted to fix it later on, but right now I have a different idea:

* drop the assertion because it was mainly supposed to be a helper to ease the migration. Currently, everyone who runs synapse should have postgres configured properly and everyone who's about to do so can see from the docs that configuring postgres manually is required, i.e. there's no real value to keep the assertion.
* add a `requires = ` that adds `postgresql.service` the same way `after` does. This will work fine for the majority of the cases.
* in your edge case you can fix the problem on your own entirely by doing `systemd.services.matrix-synapse.requires = lib.mkForce [];`.

And finally, we don't need to come up with an option name on our own then ;-)

wdyt?